internal/server/data/access_key.go

```diff
@@ -112,21 +129,33 @@ func GetAccessKey(tx GormTxn, selectors ...SelectorFunc) (*models.AccessKey, err
 	return result, nil
 }
 
-func DeleteAccessKey(db GormTxn, id uid.ID) error {
-	return delete[models.AccessKey](db, id)
+type DeleteAccessKeysOptions struct {
+	// ByID instructs DeleteAccessKeys to delete the key with this ID.
+	ByID uid.ID
+	// ByUserID instructs DeleteAccessKeys to delete keys issued for this user.
+	ByUserID uid.ID
+	// ByProviderID instructs DeleteAccessKeys to delete keys issued by this
+	// provider.
+	ByProviderID uid.ID
 }
 
-func DeleteAccessKeys(db GormTxn, selectors ...SelectorFunc) error {
-	toDelete, err := list[models.AccessKey](db, nil, selectors...)
-	if err != nil {
-		return err
-	}
-
-	ids := slice.Map[models.AccessKey, uid.ID](toDelete, func(k models.AccessKey) uid.ID {
-		return k.ID
-	})
-
-	return deleteAll[models.AccessKey](db, ByIDs(ids))
+func DeleteAccessKeys(tx WriteTxn, opts DeleteAccessKeysOptions) error {
+	query := Query("UPDATE access_keys")
+	query.B("SET deleted_at = ? WHERE", time.Now())
+	switch {
+	case opts.ByID != 0:
+		query.B("id = ?", opts.ByID)
+	case opts.ByUserID != 0:
+		query.B("issued_for = ?", opts.ByUserID)
+	case opts.ByProviderID != 0:
+		query.B("provider_id = ?", opts.ByProviderID)
+	default:
+		return fmt.Errorf("DeleteAccessKeys requires an ID to delete")
+	}
+	query.B("AND organization_id = ?", tx.OrganizationID())
```

Comment:
There isn't really any shared code anymore, which I see as advantage. The queries are very explicit and obvious now, I don't have to dig through hundreds of lines to guess at how the query is built.

I could add a panic here when `OrganizationID` is 0, but I think if we step back a bit we'll see even that is not adding anything.

This `AND organization_id` isn't necessary in this query. We're already using IDs in the where clause, and all of those IDs (primary key, user id, or provider id) can only refer to items in the same org (if they were different orgs we'd have a bug on insert).

So even removing this `AND organization_id` would be safe, and potentially makes this code easier to use (because we don't have to artificially set an organization ID in a transaction when we don't need one).

I added this extra clause as a safeguard, but I think safeguards on safeguards is going too far.